src/core/lib/gprpp/fixed_capacity_vector.h

```diff
@@ -0,0 +1,106 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_LIB_GPRPP_FIXED_CAPACITY_VECTOR_H
+#define GRPC_CORE_LIB_GPRPP_FIXED_CAPACITY_VECTOR_H
+
+#include <grpc/support/port_platform.h>
+
+#include <cassert>
+
+#include "src/core/lib/gprpp/memory.h"
+
+namespace grpc_core {
+
+// This is similar to std::vector<>, except that its capacity is fixed at
+// construction time.  This allows allocating the entire vector in a
+// single memory allocation.
+
+// TODO(roth): This interface does not currently implement all methods
+// supported by STL containers (e.g., iterators).  Those can be added as
+// needed.
+
+template <typename T>
+class FixedCapacityVector {
+ public:
+  static UniquePtr<FixedCapacityVector> Create(size_t capacity) {
+    // Allocate space for both this object and the elements.
+    FixedCapacityVector* v = static_cast<FixedCapacityVector*>(
+        gpr_malloc(sizeof(FixedCapacityVector) + (capacity * sizeof(T))));
```

Comment:
So FixedCapacityVector is likely to have size 8, which would make the actual data located here to be on an offset of 8 relative to the 16-aligned block returned by malloc. Is there any situation where the actual data array needs to be 16-aligned?

Reply:
Good point.  Fixed.